### DISCUSSION TITLE
How to invalidate scientific notation in numeric inputs?

### DISCUSSION THREAD

**Discussion by mterczynski** [react-hook-form/react-hook-form#8661] [Q&A]
Hello there,

I wanted to ask if it's possible to apply pattern validation for `type="number"` inputs in a way that would work no matter if the user typed in digits or used letters. The problem that I currently have is that `"1e"` or `"e"`, `"1e2222"` doesn't result in any errors even though they don't match `/^\d*$/` pattern. Below I've attached the full code I have used.

```typescript
import { useForm } from "react-hook-form";

export default function App() {
  const {
    register,
    formState: { errors },
  } = useForm({ mode: "all" });

  return (
    <form>
      <input
        type={"number"}
        {...register("example", {
          pattern: {
            value: /^\d*$/,
            message: "invalid number",
          },
        })}
      />

      {errors.example && <span>{(errors.example as any).message}</span>}
    </form>
  );
}
```

**Comment by bluebill1049** [comment]
i think it should, can you share a working codesandbox?

**Reply by mterczynski** [reply]
Yes Bill, there you go:
https://codesandbox.io/s/restless-cloud-9lkjhe

**Comment by Moshyfawn** [comment]
Actually, the HTML specs mention that the pattern attribute doesn't work with `type="number"`.

See [the current MDN attributes spec](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#attributes)

> | **Attribute** | **Type or Types**                           | **Description**                              |
> |-----------|-----------------------------------------|------------------------------------------|
> | [pattern](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#pattern)   | `text`, `search`, `url`, `tel`, `email`, `password` | Pattern the `value` must match to be valid |

Also, a quote from [the previous MDN input pattern attribute section version](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#attr-pattern=) _(currently unavailable)_

> **`pattern`** - A regular expression that the control's value is checked against. The pattern must match the entire value, not just some subset. Use the title attribute to describe the pattern to help the user. **This attribute applies when the value of the type attribute is text, search, tel, url or email; otherwise it is ignored**. The regular expression language is the same as JavaScript's. The pattern is not surrounded by forward slashes.

Additionally, check out [the WHATWG attributes application section to different input types](https://html.spec.whatwg.org/multipage/input.html#concept-input-apply)

> | **Content attributes** | **Hidden** | **Text, Search** | **URL, Phone** | **Email** | **Password** | **Date, Month, Week, Time** | **Local Date and Time** | **Number** | **Range** | **Color** | **Checkbox, Radio Button** | **File Upload** | **Submit Button** | **Image Button** | **Reset Button, Button** |
> |--------------------|:------:|:------------:|:----------:|:-----:|:--------:|:-----------------------:|:-------------------:|:------:|:-----:|:-----:|:----------------------:|:-----------:|:-------------:|:------------:|:--------------------:|
> | [pattern](https://html.spec.whatwg.org/multipage/input.html#attr-input-pattern)            |    ·   |      Yes     |     Yes    |  Yes  |    Yes   |            .            |          .          |    .   |   .   |   .   |            .           |      .      |       .       |       .      |           .          |

**Reply by mterczynski** [reply]
Thanks for the answer @Moshyfawn.
I have opened the element inspector and I see the `pattern` attribute is not present on the input so that's not an issue.
Please notice that we are passing `pattern` to `register` function, and not to the input directly.
